src/rules.ts

```diff
@@ -1,12 +1,68 @@
 import {
   noopTest,
   edit
-} from './helpers.js';
+} from './helpers.ts';
+import type { Rule, Rules } from './MarkedOptions.ts';
+
+type BlockRuleNames =
+    | 'newline'
+    | 'code'
+    | 'fences'
+    | 'hr'
+    | 'heading'
+    | 'blockquote'
+    | 'list'
+    | 'html'
+    | 'def'
+    | 'lheading'
+    | '_paragraph'
+    | 'text'
+    | '_label'
+    | '_title'
+    | 'bullet'
+    | 'listItemStart'
+    | '_tag'
+    | '_comment'
+    | 'paragraph'
+    | 'uote' ;
+
+type BlockSubRuleNames = 'normal' | 'gfm' | 'pedantic';
+
+type InlineRuleNames =
+    | 'escape'
+    | 'autolink'
+    | 'tag'
+    | 'link'
+    | 'reflink'
+    | 'nolink'
+    | 'reflinkSearch'
+    | 'code'
+    | 'br'
+    | 'text'
+    | '_punctuation'
+    | 'punctuation'
+    | 'blockSkip'
+    | 'escapedEmSt'
+    | '_comment'
+    | '_escapes'
+    | '_scheme'
+    | '_email'
+    | '_attribute'
+    | '_label'
+    | '_href'
+    | '_title'
+    | 'strong'
+    | '_extended_email'
+    | '_backpedal';
+
+type InlineSubRuleNames = 'gfm' | 'emStrong' | 'normal' | 'pedantic'| 'breaks';
 
 /**
  * Block-Level Grammar
  */
-export const block = {
+// Not all rules are defined in the object literal
+// @ts-expect-error
+export const block: Record<BlockRuleNames, Rule> & Record<BlockSubRuleNames, Rules> & Rules = {
```

Comment:
The part in the block object is no problem and would work nicely with this type, but the code below the initial object definition would need a typecast on every usage where the rules are passed into `edit`, `replace` or are spread into subrules.

Reply:
What is the error that happens if we don't typecast?

Reply:
There are two categories of failures:
The first one for
```
block.normal = { ...block };
```
it would complain with 
```
TS2322: Type  { [x: string]: Rule; }  is not assignable to type  Rule 
Type  { [x: string]: Rule; }  is missing the following properties from type  RegExp :  exec, test, source, global , and  14  more.
```

The second category concerns 
```
block.gfm.table = edit(block.gfm.table as Rule)
```
and similar. It complains with 
```
TS2339: Property  table  does not exist on type  Rule 
Property  table  does not exist on type  string 
```